### PR TITLE
Remove need for expression languages (use JsonPath only)

### PR DESCRIPTION
**Pull request by tsurdilo**
Signed-off-by: Tihomir Surdilovic <tsurdilo@redhat.com>

**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts this PR updates:**

- [x] Specification
- [x] Examples
- [x] Schema
- [x] Roadmap
- [ ] Use Cases
- [ ] Community
- [ ] TCK
- [ ] Other

**What this PR does / why we need it**:

Use of expression languages lessens the portability of the specification. Recently we made a change 
that offers CEL as only expression lang, but that was a mistake, as all we need are JsonPath expressions.

This change removes the need for expression langs thus increasing the spec portability.

**Special notes for reviewers**:

**Additional information (if needed):**

### CODE REVIEW

**Comment by tsurdilo** [comment]
Adds a Workflow Data Expressions which clearly defines how expressions can be defined inside string properties.

File: specification.md [hunk 1]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
 
 ## Specification Goals
 
-At the core of the Serverless Workflow specification is its [JSON Schema](schema/workflow.json).
+At the core of the Serverless Workflow specification is described via its [JSON Schema](schema/workflow.json).
 This schema defines the workflow model. It can also be used for generation of many different artifacts
-such as APIs and SPIs. We plan to provide these in the near future, and hope to expand them 
-to many different languages. This specification also strives to provide a TCK with a set of tests soon, whose 
-implementations can be used to test conformance to the specification.
+such as APIs and SPIs and SDKs. Currently the specification provides SDKs for Go and Java and we hope to
```
Comment (ricardozanini) [inline specification.md:111]:
```suggestion
such as APIs and SPIs and SDKs. Currently the specification provides SDKs for [Go](https://github.com/serverlessworkflow/sdk-go) and [Java](https://github.com/serverlessworkflow/sdk-java) and we hope to 
```

File: specification.md [hunk 2]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
-At the core of the Serverless Workflow specification is its [JSON Schema](schema/workflow.json).
+At the core of the Serverless Workflow specification is described via its [JSON Schema](schema/workflow.json).
 This schema defines the workflow model. It can also be used for generation of many different artifacts
-such as APIs and SPIs. We plan to provide these in the near future, and hope to expand them 
-to many different languages. This specification also strives to provide a TCK with a set of tests soon, whose 
-implementations can be used to test conformance to the specification.
+such as APIs and SPIs and SDKs. Currently the specification provides SDKs for Go and Java and we hope to 
+expand this list in the future. 
+
+We also strives to provide a Technology Compatibility Kit (TCK) which provides a suite of tests usec for
```
Comment (ricardozanini) [inline specification.md:114]:
```suggestion
We also strive to provide a Technology Compatibility Kit (TCK) which provides a suite of tests used for
```

File: specification.md [hunk 3]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
-### Workflow Expressions
+Workflow model parameters may use expressions to access the JSON data. 
+Note that different data filters play a big role as to which parts of the state data is queried against. Reference the 
+[State Information Filtering](#State-Information-Filtering) section for more information.
 
-Different parts of the workflow definition such as [transitions](#Transition-Definition), [retries](#Retry-Definition), [errors](#Error-Definition), 
-etc include "expression" parameters. 
-These parameters define expressions that have to be evaluated against state or error data.
+All expressions must follow the [JsonPath](https://github.com/json-path/JsonPath) format and can be evaluated with a JsonPath expression evaluator. 
+JSONPath expressions can traverse the workflow JSON data.
```
Comment (ricardozanini) [inline specification.md:167]:
```suggestion
JsonPath expressions can traverse the workflow JSON data. 
```

File: specification.md [hunk 4]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
+
+We can use an expression inside a string type parameter, for example:
+
+```json
+{
+    "paramName": "Hello {{ $.firstName }} {{ $.lastName }}"
+}
+```
+
+would set the value of "paramName" to "Hello John Doe".
```
Comment (ricardozanini) [inline specification.md:210]:
```suggestion
would set the value of `paramName` to `Hello John Doe`.
```

File: specification.md [hunk 5]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
+Some parameters require an expression to resolve to a boolean (true / false). In this case JsonPath expressions can also be used.
+The expression should evaluate to true, if the result contains a subset of the JSON data, and false if it is empty. For example:
+
+```json
+{
+   "isAdult" : "{{ $.[?(@.age  > 18)] }}"
+}
+```
+
+would set the value if "isAdult" to true, as the expression returns a non-empty subset of the JSON data. Similarly
```
Comment (ricardozanini) [inline specification.md:246]:
```suggestion
would set the value if `isAdult` to true, as the expression returns a non-empty subset of the JSON data. Similarly 
```

File: specification.md [hunk 6]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
+
+would set the value if "isAdult" to true, as the expression returns a non-empty subset of the JSON data. Similarly 
+
+```json
+{
+   "isChild" : "{{ $.[?(@.age  < 18)] }}"
+}
+```
+
+would return false as the result of the JSONPath expression is empty.
```
Comment (ricardozanini) [inline specification.md:254]:
```suggestion
would return false as the result of the JsonPath expression is empty.
```

File: specification.md [hunk 7]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
+
+JsonPath also includes a limited set of built-in functions that can be used inside expressions. For example the expression
+
+```json
+{
+   "phoneNums" : "{{ $.phoneNumbers.length() }}"
+}
+```
+
+would set the value of 'phoneNums' to "2".
```
Comment (ricardozanini) [inline specification.md:264]:
```suggestion
would set the value of `phoneNums` to `2`.
```

File: specification.md [hunk 8]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
+
+As previously mentioned, expressions are evaluated against the subset of data that can see. For example 
+the "parameters" param of the [functionRef definition](#FunctionRef-Definition) can evaluate expressions
```
Comment (ricardozanini) [inline specification.md:267]:
```suggestion
the `parameters` param of the [functionRef definition](#FunctionRef-Definition) can evaluate expressions 
```

File: specification.md [hunk 9]
```diff
@@ -106,51 +106,187 @@ incoming events can trigger function invocations during workflow execution.
...
+
+```json
+{
+   "id": "MySampleWorkflow",
+   "name": "Sample Workflow",
+   "version": "{{ $.inputVersion }}"
+}
+```
+
+which would set the workflow version to "1.0.0".
```
Comment (ricardozanini) [inline specification.md:289]:
```suggestion
which would set the workflow version to `1.0.0`.
```

File: specification.md (hunk 1 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 2 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 4 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 3 above)
Comment (tsurdilo) [reply specification.md]: fixed and also updated all places to change from "JSONPath" to "JsonPath"

File: specification.md (hunk 5 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 7 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 8 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 6 above)
Comment (tsurdilo) [reply specification.md]: fixed

File: specification.md (hunk 9 above)
Comment (tsurdilo) [reply specification.md]: fixed

**Comment by tsurdilo** [comment]
@ricardozanini thanks for review! Updated with all your suggestions.

**Pull request merged**